src/tir/transforms/hoist_if_then_else.cc

```diff
@@ -0,0 +1,376 @@
+ *                A[3*i+2j+k] = B[7*i+3j+k]
+ *
+ * Only the For with k variable will be considered and the resulting stmt would be:
+ * for (i = 0; i < 3; i++)
+ *    for (j = 0; j < 4; j++)
+ *        A[i + j] = A[i + j] - 1
+ *        if (likely(i*2 < 4))
+ *            for (k = 0; k < 5; k++)
+ *                A[3*i+2j+k] = B[7*i+3j+k]
+ *
+ * This pass doesn't do hoisting for consecutive IfThenElse stmt. The following
+ * block won't be optimized:
+ * for (i = 0; i < 3; i++)
+ *    for (j = 0; j < 4; j++)
+ *        for (k = 0; k < 5; k++)
+ *            if (likely(i*2 < 4))
+ *                A[3*i+2j+k] = B[7*i+3j+k]
+ *            if (likely(j > 2))
+ *                A[i+j+k] = B[i+j+k]
+ *
+ */
+
+// Select potential candidate IRs that can be hoisted.
+class HoistCandidateSelector final : public StmtExprVisitor {
+ public:
+  HoistCandidateSelector() { InitRecorder(); }
+
+  void VisitStmt_(const ForNode* op) final {
+    // Check if it is first for loop, then start the recorder
+    if (!RecordingComplete()) {
+      StartOrAddRecord(op);
+      StmtExprVisitor::VisitStmt_(op);
+      RemoveRecord(op);
+      return;
+    }
+
+    StmtExprVisitor::VisitStmt_(op);
+  }
+
+  void VisitStmt_(const SeqStmtNode* op) final {
+    // If SeqStmt is encountered in the middle of recording
+    //  then need to purge all, as it can not be hoisted
+    if (IsRecordingOn()) {
+      ResetRecorder();
+    }
+    StmtExprVisitor::VisitStmt_(op);
+  }
+
+  void VisitStmt_(const AttrStmtNode* op) final {
+    // Maintain list of all vars in AttrStmt
+    // To stop hoisting if any of the block variables are used.
+    //
+    // NOTE: If in future
+    // hoisting is required for any specific case,
+    // then add exception to only those case
+    // rather than allowing for all.
+    UpdateAttrVarList(op);
+    StmtExprVisitor::VisitStmt_(op);
+    RemoveAttrVarList(op);
+  }
+
+  void VisitStmt_(const IfThenElseNode* op) final {
+    if (IsRecordingOn()) {
+      is_if_cond = true;
+      StmtExprVisitor::VisitExpr(op->condition);
+      is_if_cond = false;
+
+      if (CheckValidIf()) {
+        // Check corresponding for loop
+        bool match_found = false;
+        size_t match_for_loop_pos = 0;
+        for (auto var : if_var_list_) {
+          for (size_t i = 0; i < ordered_for_list_.size() - 1; ++i) {
+            if (ordered_for_list_[i] == var_for_map_[var]) {
+              if (match_for_loop_pos < i) {
+                match_for_loop_pos = i;
+              }
+              match_found = true;
+              break;
+            }
+          }
+        }
+        // If none of the for loop has the matching loop variable as if condition,
+        // then the if node need to be hoisted on top of all, provided no parent loop exists.
+        int target_for_pos = match_found ? match_for_loop_pos + 1 : 0;
+
+        // Check if target for loop is not the parent of current if node
+        if (!IsParentForLoop(target_for_pos)) {
```

Comment:
I am sorry for the confusion!
We can definitely add more comments to it. Let me put an explanation to the logic, then we can figure out the necessary comment for it together.

Above case will hit in 2 cases:
case 1(Partial Match found): 
case 2 (No Match found) :

Reply:
In both cases above, `target_for_pos` ends up to be the inner-most loop, i.e., equals to `n`. I think we only have to check whether `target_for_pos < n`, instead of running `IsParentForLoop`. Am I right?

Reply:
Sorry the `target_for_pos` values will be different in both cases. 
For partial match or exact match, it can be [0, n-1] and if no match found it will be 0. To accommodate all the scenarios, hence the common logic :)

Reply:
Let me describe my understanding.

In case 1, `n` is 2, where Loop `i`, `j`, `k` is Loop 0, 1, 2, respectively. Both Loop 0 and 1 match, so `match_for_loop_pos == 1`, and `target_for_pos == match_for_loop_pas + 1 == 2`, which is equal to `n`.

In case 2, `n` is 0, where Loop `k` is Loop 0, while Loop `i` or `j` is not recorded. There is no match, so `target_for_pos == 0`, which is still equal to `n`.

What is wrong?

Reply:
May be the example i shares are not enough to cover all cases :)
Let me put another example, may be that will clear-up all the queries.
```
for (i: int32, 0, l: int32) {
  for (j: int32, 0, m: int32) {
    for (k: int32, 0, n: int32) {
      if @tir.likely(tvm.tir.any(i < 4, k >= 8)), dtype=bool) {
        m
      } else {
        n
      }
    }
  }
}
```
In the case above, `target_for_pos = 1`, but we should not do hoisting in that case. 
I hope this will clear your queries.

Reply:
Why is `target_for_pos = 1`?

```c++
if (match_for_loop_pos < i) {
  match_for_loop_pos = i;
}
```

This update hits twice, once for `i` and one for `k`, and `match_for_loop_pos` ends up to be 2, so `target_for_pos == 3`.

Reply:
The loop above runs only n-2 times, in this case it is 1 time, that's why the value for target_for_pos = 1.

Reply:
Let me make it explicit.

```
if_var_list_ = [i, k]
ordered_for_list_ = [for i, for j, for k]
Iter 1: var == i, i == 0. (ordered_for_list_[i] == var_for_map_[var]) == true; match_for_loop_pos = 0; break;
Iter 2: var == k, i == 0. (ordered_for_list_[i] == var_for_map_[var]) == false;
Iter 3: var == k, i == 1. (ordered_for_list_[i] == var_for_map_[var]) == false;
Iter 4: var == k, i == 2. (ordered_for_list_[i] == var_for_map_[var]) == true; match_for_loop_pos = 2; break;
target_for_pos = match_for_loop_pos + 1 = 3
```

Reply:
Ah! I understand now. Why the confusion. Sorry if my explanation was poor earlier. 

Actually in your assumption above :
It won't iterate till step 3, execute till step 2 only. 

Because the for loop cond is
( i < ordered_for_list_.size() - 1)

Reply:
Oh, got it, but why? Why not change it to `(i < ordered_for_list_.size())` so we don't need to check `IsParentForLoop`?

Reply:
Yes, you are right, for some cases we dont have to check `IsParentForLoop` 
if we traverse `(i < ordered_for_list_.size())`.
But i am afraid the logic wont be a generalized one. Still there will be cases when we need to check it. 
And above all it is a fail safe operation, which will never allow the hoisting to cross its parent for node.

Reply:
That's fine. We can merge this PR first.